### PR TITLE
Fail without enqueueing iff the airbyte message type is unrecognized

### PR DESCRIPTION
**Pull request by johnny-schmidt**
## What
Addresses [this Zenhub issue](https://app.zenhub.com/workspaces/destinations-6332532d6893ce001c7a56a7/issues/gh/airbytehq/airbyte-internal-issues/8064): when the airbyte message type is unrecognized, log an error instead of throwing the exception.

## How
1. Error handling/obfuscation is lifted out of the json deserializer (which is type-agnostic) into the caller (AirbyteMessageDeserializer)
2. The caller's return value is made nullable
3. The caller detects the specific failure (top-level enum type only), logs the unrecognized type only (no other client data context) and returns null
4. The caller's caller skips enqueuing on null

NOTE: This means that we will never call enqueue with the byte size passed into accept. However, @stephane-airbyte noticed that we do nothing with this when the type is unrecognized anyway (which may be a mistake if something's relying on that matching the calls to accept).

cc @edgao 

## User Impact
User should notice very little. In the rare event we add a new message type, it will not be a breaking change.

## Can this PR be safely reverted and rolled back?
- [ X] YES 💚
- [ ] NO ❌
Yes, as long as we are not in the middle of the process of releasing a new message type. :)

### CODE REVIEW

File: airbyte-cdk/java/airbyte-cdk/core/src/test/kotlin/io/airbyte/cdk/integrations/base/AirbyteTraceMessageUtilityTest.kt [hunk 1]
```diff
@@ -48,6 +48,7 @@ class AirbyteTraceMessageUtilityTest {
             Mockito.mock(RuntimeException::class.java),
             "this is a config error"
         )
+        System.out.println("HERE:" + outContent.toString(StandardCharsets.UTF_8))
```
Comment (edgao) [inline airbyte-cdk/java/airbyte-cdk/core/src/test/kotlin/io/airbyte/cdk/integrations/base/AirbyteTraceMessageUtilityTest.kt:51]: nit: remove this

File: airbyte-cdk/java/airbyte-cdk/core/src/test/kotlin/io/airbyte/cdk/integrations/destination/async/AsyncStreamConsumerTest.kt [hunk 2]
```diff
@@ -732,4 +733,38 @@ class AsyncStreamConsumerTest {
             }
         assertEquals(expRecords, actualRecords)
     }
+
+    @Test
+    internal fun deserializeAirbyteMessageWithUnrecognizedType() {
+        val airbyteMessage = AirbyteMessage().withType(AirbyteMessage.Type.RECORD)
+        val serialized = Jsons.serialize(airbyteMessage)
```
Comment (edgao) [inline airbyte-cdk/java/airbyte-cdk/core/src/test/kotlin/io/airbyte/cdk/integrations/destination/async/AsyncStreamConsumerTest.kt:740]: why not just directly hardcode the json string? I feel like going through the message -> serialize -> replace makes this test less clear
```kotlin
val retyped = """{"type": "__UNKNOWN_TYPE_OF_MESSAGE__"}"""
```

File: airbyte-cdk/java/airbyte-cdk/dependencies/src/main/kotlin/io/airbyte/commons/json/Jsons.kt [hunk 3]
```diff
@@ -139,6 +139,11 @@ object Jsons {
         }
     }
 
+    @JvmStatic
```
Comment (edgao) [inline airbyte-cdk/java/airbyte-cdk/dependencies/src/main/kotlin/io/airbyte/commons/json/Jsons.kt:142]:
```suggestion
    /**
     * CAUTION: Using this method may leak the raw json string to stdout via a thrown exception.
     * Callers MUST either verify that the json string does not contain sensitive information,
     * or make sure to catch the exception and use [obfuscate] to strip out potentially-sensitive information.
     */
    @JvmStatic
```

**Review by edgao — APPROVED** [review]
couple nitpicks, LGTM once those are resolved

FYI how to publish the cdk:
* https://github.com/airbytehq/airbyte/blob/master/airbyte-cdk/java/airbyte-cdk/README.md#bumping-the-cdk-version
* and also add an entry to the changelog at the bottom

File: airbyte-cdk/java/airbyte-cdk/core/src/test/kotlin/io/airbyte/cdk/integrations/destination/async/AsyncStreamConsumerTest.kt (hunk 2 above)
Comment (johnny-schmidt) [reply airbyte-cdk/java/airbyte-cdk/core/src/test/kotlin/io/airbyte/cdk/integrations/destination/async/AsyncStreamConsumerTest.kt:740]: Technically I don't know if the builder is adding any other metadata to make it a legal message?

File: airbyte-cdk/java/airbyte-cdk/dependencies/src/main/kotlin/io/airbyte/commons/json/Jsons.kt (hunk 3 above)
Comment (johnny-schmidt) [reply airbyte-cdk/java/airbyte-cdk/dependencies/src/main/kotlin/io/airbyte/commons/json/Jsons.kt:142]: Good suggestion!

File: airbyte-cdk/java/airbyte-cdk/core/src/test/kotlin/io/airbyte/cdk/integrations/destination/async/AsyncStreamConsumerTest.kt (hunk 2 above)
Comment (stephane-airbyte) [reply airbyte-cdk/java/airbyte-cdk/core/src/test/kotlin/io/airbyte/cdk/integrations/destination/async/AsyncStreamConsumerTest.kt:740]: I think either way is fine. The error and catching is explicit enough that only the relevant issues get filtered out

File: airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/integrations/destination/async/deser/AirbyteMessageDeserializer.kt [hunk 4]
```diff
@@ -3,37 +3,65 @@
...
+                    // is generated from json-schema.
+                    val pat =
+                        Regex(
+                            "Cannot construct instance of .*AirbyteMessage.Type., problem: ([_A-Z]+)"
+                        )
+                    val match = pat.find(e.message!!)
+                    if (match != null) {
+                        val unrecognized = match.groups[1]?.value
+                        logger.warn { "Unrecognized message type: $unrecognized" }
+                        return null
```
Comment (stephane-airbyte) [inline airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/integrations/destination/async/deser/AirbyteMessageDeserializer.kt:57]: I'd rather throw a typed exception than return null here. This way, we also avoid all the !! everywhere we know this to be valid

Comment (johnny-schmidt) [reply airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/integrations/destination/async/deser/AirbyteMessageDeserializer.kt:57]: Done.

**Comment by johnny-schmidt** [comment]
@stephane-airbyte I think your comments have been addressed.

**Review by stephane-airbyte — APPROVED** [review]

**Pull request merged**